File: src/services/notification.ts
//#==IqwanEngine: Telegram Notification Service
import dotenv from 'dotenv';
dotenv.config();

const BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const ADMIN_ID = process.env.ADMIN_ID;

/**
 * Sends a Telegram notification to the admin.
 * @param message The text to send
 * @param buttons Optional inline keyboard buttons
 */
export async function sendTelegramNotification(message: string, buttons: any[] = []) {
  if (!BOT_TOKEN || !ADMIN_ID) {
    console.warn("//#==IqwanEngine: Telegram credentials missing. Skipping notification.");
    return;
  }

  const url = `https://api.telegram.org/bot${BOT_TOKEN}/sendMessage`;
  
  const payload: any = {
    chat_id: ADMIN_ID,
    text: message,
    parse_mode: 'HTML',
  };

  if (buttons.length > 0) {
    payload.reply_markup = {
      inline_keyboard: [buttons]
    };
  }

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error("//#==IqwanEngine: Telegram API Error:", errorText);
    }
  } catch (error) {
    console.error("//#==IqwanEngine: Failed to send Telegram notification:", error);
  }
}

/**
 * Tier 1: Visitor Alert
 */
export async function notifyVisit(companyName: string) {
  const message = `<b>IqwanEngine Notification</b>\n\nIqwan, ada orang melawat Recruiter AI anda dari <b>${companyName}</b>.`;
  const buttons = [
    { text: "🔍 Lihat Info Syarikat", url: `https://www.google.com/search?q=${encodeURIComponent(companyName)}` }
  ];
  return sendTelegramNotification(message, buttons);
}

/**
 * Tier 2: Interest Alert
 */
export async function notifyInterest(companyName: string, details: string) {
  const message = `<b>🚨 HIGH-VALUE ALERT</b>\n\nIqwan, ada orang berminat dengan anda, mungkin boleh tackle dia.\n\n<b>Company:</b> ${companyName}\n<b>Details:</b> ${details}`;
  const buttons = [
    { text: "📱 Details Recruiter", callback_data: "view_details" }, // Note: handling callbacks requires a separate listener
    { text: "🔍 Lihat Info Syarikat", url: `https://www.google.com/search?q=${encodeURIComponent(companyName)}` }
  ];
  return sendTelegramNotification(message, buttons);
}

File: api/services/notification.py
import logging
import os

import requests
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger("IqwanEngine.TelegramService")
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)


class TelegramService:
    def __init__(self):
        # Mandatory Security: Load credentials from environment
        self.token = os.getenv("TELEGRAM_BOT_TOKEN")
        self.admin_id = os.getenv("ADMIN_ID")
        self.base_url = (
            f"https://api.telegram.org/bot{self.token}/sendMessage"
            if self.token
            else None
        )

    def send_notification(self, message, buttons=None):
        """Helper to send telegram messages with optional buttons."""
        if not self.token or not self.admin_id:
            logger.warning("IqwanEngine: Skipping Telegram - Credentials missing.")
            return False

        if not self.base_url:
            logger.error("IqwanEngine: Telegram base URL not configured.")
            return False

        payload = {"chat_id": self.admin_id, "text": message, "parse_mode": "HTML"}

        if buttons:
            payload["reply_markup"] = {"inline_keyboard": [buttons]}

        try:
            response = requests.post(self.base_url, json=payload)
            response.raise_for_status()  # Raise HTTPError for bad responses (4xx or 5xx)
            return response.ok
        except requests.exceptions.RequestException as e:
            logger.error("IqwanEngine: Telegram Webhook Error: %s", e)
            return False
        except Exception as e:
            logger.error("IqwanEngine: Unexpected error in TelegramService: %s", e)
            return False

    def notify_visit(self, company_name):
        """Tier 1 Notification: Visitor Alert"""
        message = (
            f"✨ <b>[ IQWANENGINE : VISIT ALERT ]</b> ✨\n"
            f"━━━━━━━━━━━━━━━━━━━━━\n\n"
            f"🏢 <b>Company:</b>\n"
            f"┗ <code>{company_name}</code>\n\n"
            f"━━━━━━━━━━━━━━━━━━━━━\n"
            f"👀 <i>Iqwan, someone just visited your Recruiter AI. Go check it out!</i> 🚀"
        )
        buttons = [
            {
                "text": "🔍 Search Company Info",
                "url": f"https://www.google.com/search?q={company_name}",
            }
        ]
        return self.send_notification(message, buttons)

    def notify_interest(self, company_name, details):
        """Tier 2 Notification: High-Value Lead Alert"""
        message = (
            f"🔥 <b>[ IQWANENGINE : HOT INTEREST ]</b> 🔥\n"
            f"━━━━━━━━━━━━━━━━━━━━━\n\n"
            f"🏢 <b>Company:</b>\n"
            f"┗ <code>{company_name}</code>\n\n"
            f"📋 <b>Additional Details:</b>\n"
            f"┗ <code>{details}</code>\n\n"
            f"━━━━━━━━━━━━━━━━━━━━━\n"
            f"🎯 <i>Iqwan, a prospect is interested in you! Great opportunity to pitch them.</i> ⚡"
        )
        buttons = [
            {
                "text": "📄 Send Resume",
                "url": "https://your-admin-dashboard.com",
            },
            {
                "text": "🔍 Search Company Info",
                "url": f"https://www.google.com/search?q={company_name}",
            },
        ]
        return self.send_notification(message, buttons)


telegram_service = TelegramService()
